### PR TITLE
Add tests for UsersManager

### PR DESCRIPTION
**Pull request by fflorent**
## Context

HomeDBManager lacks of direct tests, which makes hard to make rework or refactorations.

## Proposed solution

Specifically here, I introduce tests which call exposed UsersManager methods directly and check their result.

Also:
 - I removed `updateUserName` which seems to me useless (`updateUser` does the same work)
 - Taking a look at the `getUserByLogin` methods, it appears that Typescirpt infers it returns a `Promise<User|null>` while in no case it may resolve a nullish value, therefore I have forced to return a `Promise<User>` and have changed the call sites to reflect the change. 

## Related issues

I make this change for then working on #870 

## Has this been tested?

- [x] 👍 yes, I added tests to the test suite
- [ ] 💭 no, because this PR is a draft and still needs work
- [ ] 🙅 no, because this is not relevant here
- [ ] 🙋 no, because I need help

### CODE REVIEW

**Review by fflorent — COMMENTED** [review]
Comments after having taken a look at the PR with @hexaltation

**Review by hexaltation — APPROVED** [review]
LGTM

File: app/server/companion.ts [hunk 1]
```diff
@@ -194,7 +190,7 @@ export function addDbCommand(program: commander.Command,
       if (!process.env.TYPEORM_LOGGING) {
         process.env.TYPEORM_LOGGING = 'true';
       }
-      const connection = reuseConnection || await getOrCreateConnection();
+      const connection = reuseConnection || await createNewConnection();
```
Comment (berhalak) [inline app/server/companion.ts:193]: Can you add a comment why this was changed? The `reuseConnection` might be a different one from that returned by `getOrCreateConnection`.

**Review by berhalak — CHANGES_REQUESTED** [review]

File: test/gen-server/lib/homedb/UsersManager.ts [hunk 2]
```diff
@@ -0,0 +1,996 @@
...
+import { ANONYMOUS_USER_EMAIL, EVERYONE_EMAIL, PREVIEWER_EMAIL, UserOptions } from 'app/common/UserAPI';
+import { Login } from 'app/gen-server/entity/Login';
+import { Pref } from 'app/gen-server/entity/Pref';
+import { EntityManager } from 'typeorm';
+import log from 'app/server/lib/log';
+import winston from 'winston';
+import { updateDb } from 'app/server/lib/dbUtils';
+
+const username = process.env.USER || "nobody";
+const tmpDir = path.join(tmpdir(), `grist_test_${username}_userendpoint`);
```
Comment (berhalak) [inline test/gen-server/lib/homedb/UsersManager.ts:28]: Should we randomize it more? Runing this test twice will reuse the same folder?
https://dev.to/omardulaimi/how-to-create-unique-temporary-directories-in-nodejs-1n69

File: test/gen-server/lib/homedb/UsersManager.ts [hunk 3]
```diff
@@ -0,0 +1,996 @@
...
+
+describe('UsersManager', function () {
+  this.timeout(30000);
```
Comment (berhalak) [inline test/gen-server/lib/homedb/UsersManager.ts:31]:
```suggestion
  this.timeout('30s');
```

File: test/gen-server/lib/homedb/UsersManager.ts [hunk 4]
```diff
@@ -0,0 +1,996 @@
...
+
+  describe('static method', function () {
+    function* makeIdxIterator() {
```
Comment (berhalak) [inline test/gen-server/lib/homedb/UsersManager.ts:34]: Can you move helper methods to the end of the file? (And even outside of the `describe` funcition)

File: test/gen-server/lib/homedb/UsersManager.ts [hunk 5]
```diff
@@ -0,0 +1,996 @@
...
+      for (let i = 0; i < 500; i++) {
+        yield i;
+      }
+    }
+
+    function makeUsers(nbUsers: number, idxIt = makeIdxIterator()) {
```
Comment (berhalak) [inline test/gen-server/lib/homedb/UsersManager.ts:40]: Can you use here the `range` function from `lodash`? I think it will look much more simpler, this one is hard to follow.

File: test/gen-server/lib/homedb/UsersManager.ts [hunk 6]
```diff
@@ -0,0 +1,996 @@
...
+        if (index.done) {
+          throw new Error('Excessive number of users created');
+        }
+        user.id = index.value;
+        user.name = `User ${index.value}`;
+        return user;
+      });
+    }
+
+    function makeUsersList(usersListLength: number, nbUsersByGroups: number) {
```
Comment (berhalak) [inline test/gen-server/lib/homedb/UsersManager.ts:53]: What does `nb` stands for?

File: test/gen-server/lib/homedb/UsersManager.ts (hunk 5 above)
Comment (fflorent) [reply test/gen-server/lib/homedb/UsersManager.ts:40]: I need to keep track of the increment for users' id across `makeUsers` call. When I call `makeUsers` let's say this way:
```ts
        const idxIt = makeIdxIterator();
        const groupsUsersMap = {
          'editors': makeUsers(2, idxIt),
          'owners': makeUsers(1, idxIt),
        };
```

groupsUsersMap equals to the following object, whose users' id are unique:
```ts
        {
          'editors': [{id: 0, name: 'user 0'}, {id: 1, name: 'user 1'}],
          'owners': [{id: 2, name: 'user 2'}]
        }
```

I don't know how to make a similar thing with `lodash.range()`.

An alternative I can do, but I find it less elegant, would consist in passing the start index each time:
```ts
let startIdx = 0;
const groupsUsersMap = {};
groupsUsersMap['editors'] = makeUsers(2, startIdx);
startIdx += 2;
groupsUsersMap['owners'] = makeUsers(1, startIdx);
startIdx += 1;
```

File: test/gen-server/lib/homedb/UsersManager.ts (hunk 2 above)
Comment (berhalak) [reply test/gen-server/lib/homedb/UsersManager.ts:28]: I'm just worry about running this tests twice (simultaneously) on the same machine, most of our tests (if not all of them) can be run simultaneously without a problem.

File: test/gen-server/lib/homedb/UsersManager.ts (hunk 4 above)
Comment (berhalak) [reply test/gen-server/lib/homedb/UsersManager.ts:37]: Ok, I don't have a strong opinion here. I just prefer to have them hidden, or not to have them at all :).

File: test/gen-server/lib/homedb/UsersManager.ts (hunk 2 above)
Comment (fflorent) [reply test/gen-server/lib/homedb/UsersManager.ts]: @berhalak I propose something to address your concerns, and still have the ability to disable the tests cleanup for debugging: 40312577e2cf761920b277e6a84d999816d02f51

What do you think?

**Comment by fflorent** [comment]
@berhalak I think of something: my test assume that they run using Sqlite.

Several questions come to my mind:
 - Can running these tests with Postgresql be worth?
 - If so: I don't see tests running with Postgresql on the CI, I can take a look to see how hard it is to support this database.
 - If not: I propose to skip these tests when `TYPEORM_TYPE` is set to `postgresql`

To the first question, my answer would be no, the UsersManager uses a lot of TypeORM methods, and does not seem to make SQL query directly  (but I may be a bit biased by my lazyness).

File: test/gen-server/seed.ts [hunk 7]
```diff
@@ -526,17 +528,28 @@ class Seed {
...
-    // There is still no official way to delete connections that I've found.
-    (getConnectionManager() as any).connectionMap = new Map();
+    await connections[0].destroy();
+    dereferenceConnection(getConnectionName());
   }
 }
 
+export function dereferenceConnection(name: string) {
+  // There seem to be no official way to delete connections.
+  // Also we should probably get rid of the use of connectionManager, which is deprecated
```
Comment (fflorent) [inline test/gen-server/seed.ts:544]: I am tempted to do that right now, but I fear that the PR becomes huge, especially to review it.

The new way consist in using `DataSource` (which is the new name for the `Connection` class, renamed [because it is confusing](https://github.com/typeorm/typeorm/issues/8010)), and instantiate using the constructor in place of `createConnection`.

We won't have any name collision, as only the ConnectionManager use it for other purposes than identifying the DataSource in error messages.

**Review by berhalak — APPROVED** [review]
Thanks @fflorent!

**Comment by berhalak** [comment]
> * Can running these tests with Postgresql be worth?

We run some tests with Postgresql in our internal CI, but since it is not part of the `grist-core` yet, and those tests are rather unit tests, I think it is fine to rely on `typeorm` abstraction.

File: test/gen-server/seed.ts [hunk 8]
```diff
@@ -526,17 +528,28 @@ class Seed {
 
 // When running mocha on several test files at once, we need to reset our database connection
 // if it exists.  This is a little ugly since it is stored globally.
-export async function removeConnection() {
-  if (getConnectionManager().connections.length > 0) {
-    if (getConnectionManager().connections.length > 1) {
+export async function removeConnection(name?: string) {
```
[the remark this answers is not in the document]
Comment (fflorent) [reply test/gen-server/seed.ts]: That's true, thanks!

**Comment by paulfitz** [comment]
I'm seeing failures in tests after UsersManager in a derived version of Grist (the one for our SaaS). Haven't looked into why yet, and may be specific to us.

Screenshot from 2024-08-23 12-19-34

File: test/gen-server/seed.ts [hunk 9]
```diff
@@ -527,16 +529,27 @@ class Seed {
...
 
+export function dereferenceConnection(name: string) {
+  // There seem to be no official way to delete connections.
+  // Also we should probably get rid of the use of connectionManager, which is deprecated
+  const connectionMgr = getConnectionManager();
+  const connectionMap = (connectionMgr as any).connectionMap as Map<string, Connection>;
+  if (!connectionMap.has(name)) {
+    throw new Error('connection with this name not found: ' + name);
+  }
+  connectionMap.delete(name);
```
Comment (paulfitz) [inline test/gen-server/seed.ts:550]: Is it possible this results in the change of behavior we see in our tests @fflorent? This is a pretty murky corner of TypeORM.

Comment (fflorent) [reply test/gen-server/seed.ts:550]: I feel more like the issue is due to this helper:
https://github.com/gristlabs/grist-core/pull/1149/files#diff-3d1ec6a021479fb3895f89f5c855b14f664e45c31e979082da277538b906cc5cR264-R282

Which has been introduced to test functions related to special ids, but I feel like these two tests bring more pain than are actually helpful:
  - [The test that check the getters related to the special IDs without initialization fail](https://github.com/gristlabs/grist-core/pull/1149/files#diff-3d1ec6a021479fb3895f89f5c855b14f664e45c31e979082da277538b906cc5cR345-R355)
  - [The test of initializeSpecialIds](https://github.com/gristlabs/grist-core/pull/1149/files#diff-3d1ec6a021479fb3895f89f5c855b14f664e45c31e979082da277538b906cc5cR957-R979)

I also tried to see whether removing the Connection related methods in favor of using DataSource would help, but with no luck (see #1166, which I may continue to work on but rather for the satisfaction of stop using deprecated entities and functions).

So I won't mind removing these tests, especially if it solves your failures, I agree that it deals with strange behaviors of TypeORM.

Comment (paulfitz) [reply test/gen-server/seed.ts:550]: It is awkward getting this sorted out, so I'm going to land and that'll force us to sort it out internally. Thanks for the suggestions.

**Comment by fflorent** [comment]
Thank you @paulfitz!
